File: agenkit-scala/src/main/scala/io/agenkit/protocols/mcp/StdioClient.scala
package io.agenkit.protocols.mcp

import java.io.*
import java.util.concurrent.atomic.AtomicLong
import java.util.concurrent.locks.ReentrantLock
import upickle.default.*

/** MCP client that communicates with a subprocess over stdin/stdout. */
class StdioClient(command: String, args: String*) extends McpClient:
  private val nextId      = AtomicLong(0)
  private val lock        = ReentrantLock()
  private var process: Process | Null = null
  private var reader: BufferedReader | Null = null
  private var writer: PrintWriter | Null = null
  private var _serverInfo = McpServerInfo()

  def initialize(): Unit =
    val cmdList = java.util.Arrays.asList((command +: args.toSeq)*)
    val pb      = ProcessBuilder(cmdList)
    pb.redirectErrorStream(false)
    process = pb.start()
    reader = BufferedReader(InputStreamReader(process.nn.getInputStream))
    writer = PrintWriter(OutputStreamWriter(process.nn.getOutputStream), true)

    val params = ujson.Obj(
      "protocolVersion" -> ProtocolVersion,
      "capabilities"    -> ujson.Obj(),
      "clientInfo"      -> ujson.Obj("name" -> "agenkit", "version" -> ClientVersion)
    )
    val resp = sendRequest("initialize", Some(params))
    resp.error.foreach(e => throw RuntimeException(s"mcp initialize error ${e.code}: ${e.message}"))
    resp.result.foreach { r =>
      r.obj.get("serverInfo").foreach { info =>
        _serverInfo = McpServerInfo(
          info.obj.get("name").map(_.str).getOrElse(""),
          info.obj.get("version").map(_.str).getOrElse("")
        )
      }
    }

  def listTools(): List[McpTool] =
    val resp = sendRequest("tools/list", None)
    resp.error.foreach(e => throw RuntimeException(s"mcp tools/list error ${e.code}: ${e.message}"))
    resp.result match
      case Some(r) =>
        r.obj.get("tools") match
          case Some(arr) => arr.arr.toList.map(v => read[McpTool](v))
          case None      => Nil
      case None => Nil

  def callTool(name: String, args: Map[String, Any]): McpToolResult =
    val argsJson = ujson.Obj.from(args.view.mapValues(anyToUjson))
    val params   = ujson.Obj("name" -> name, "arguments" -> argsJson)
    val resp     = sendRequest("tools/call", Some(params))
    resp.error.foreach(e => throw RuntimeException(s"mcp tools/call error ${e.code}: ${e.message}"))
    resp.result match
      case Some(r) => read[McpToolResult](r)
      case None    => McpToolResult(Nil, isError = true)

  def serverInfo(): McpServerInfo = _serverInfo

  def close(): Unit =
    writer = null
    reader = null
    process.nn.destroy()
    process = null

  private def sendRequest(method: String, params: Option[ujson.Value]): JsonRpcResponse =
    lock.lock()
    try
      val id  = nextId.incrementAndGet()
      val req = JsonRpcRequest("2.0", id, method, params)
      writer.nn.println(write(req))
      val line = reader.nn.readLine()
      if line == null then throw RuntimeException("mcp: server closed stdout")
      read[JsonRpcResponse](line)
    finally lock.unlock()

  private def anyToUjson(v: Any): ujson.Value = v match
    case s: String  => ujson.Str(s)
    case n: Int     => ujson.Num(n.toDouble)
    case n: Long    => ujson.Num(n.toDouble)
    case n: Double  => ujson.Num(n)
    case n: Float   => ujson.Num(n.toDouble)
    case b: Boolean => ujson.Bool(b)
    case null       => ujson.Null
    case m: Map[?, ?] =>
      ujson.Obj.from(m.asInstanceOf[Map[String, Any]].view.mapValues(anyToUjson))
    case l: List[?] =>
      ujson.Arr.from(l.asInstanceOf[List[Any]].map(anyToUjson))
    case other => ujson.Str(other.toString)

File: agenkit-java/src/main/java/io/agenkit/protocols/mcp/StdioClient.java
package io.agenkit.protocols.mcp;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicLong;
import java.util.concurrent.locks.ReentrantLock;

/**
 * MCP client that communicates with a server subprocess over stdin/stdout.
 *
 * <p>All wire I/O is line-delimited JSON-RPC 2.0. Access is serialised with a
 * {@link ReentrantLock} so that callers from different threads are safe.
 */
public class StdioClient implements McpClient {

    private final String command;
    private final List<String> args;

    private Process process;
    private BufferedReader reader;
    private BufferedWriter writer;

    private final AtomicLong nextId = new AtomicLong(0);
    private final ReentrantLock lock = new ReentrantLock();
    private final ObjectMapper mapper = new ObjectMapper();

    private McpServerInfo serverInfo = McpServerInfo.empty();

    public StdioClient(String command, String... args) {
        this.command = command;
        this.args = Arrays.asList(args);
    }

    @Override
    public void initialize() throws Exception {
        ProcessBuilder pb = new ProcessBuilder();
        List<String> cmd = new ArrayList<>();
        cmd.add(command);
        cmd.addAll(args);
        pb.command(cmd);
        pb.redirectErrorStream(false);
        process = pb.start();
        reader = new BufferedReader(new InputStreamReader(process.getInputStream()));
        writer = new BufferedWriter(new OutputStreamWriter(process.getOutputStream()));

        ObjectNode clientInfo = mapper.createObjectNode()
                .put("name", McpConstants.CLIENT_NAME)
                .put("version", McpConstants.CLIENT_VERSION);

        ObjectNode params = mapper.createObjectNode()
                .put("protocolVersion", McpConstants.PROTOCOL_VERSION);
        params.set("capabilities", mapper.createObjectNode());
        params.set("clientInfo", clientInfo);

        JsonRpcResponse resp = sendRequest("initialize", params);
        if (resp.error() != null) {
            throw new RuntimeException("mcp initialize error: " + resp.error().message());
        }
        if (resp.result() != null && resp.result().has("serverInfo")) {
            JsonNode info = resp.result().get("serverInfo");
            serverInfo = new McpServerInfo(
                    info.path("name").asText(""),
                    info.path("version").asText(""));
        }
    }

    @Override
    public List<McpTool> listTools() throws Exception {
        JsonRpcResponse resp = sendRequest("tools/list", mapper.createObjectNode());
        if (resp.error() != null) {
            throw new RuntimeException("mcp tools/list error: " + resp.error().message());
        }
        List<McpTool> tools = new ArrayList<>();
        JsonNode toolsNode = resp.result().path("tools");
        if (toolsNode.isArray()) {
            for (JsonNode t : toolsNode) {
                tools.add(mapper.treeToValue(t, McpTool.class));
            }
        }
        return tools;
    }

    @Override
    public McpToolResult callTool(String name, Map<String, Object> args) throws Exception {
        ObjectNode params = mapper.createObjectNode().put("name", name);
        params.set("arguments", mapper.valueToTree(args));
        JsonRpcResponse resp = sendRequest("tools/call", params);
        if (resp.error() != null) {
            throw new RuntimeException("mcp tools/call error: " + resp.error().message());
        }
        return mapper.treeToValue(resp.result(), McpToolResult.class);
    }

    @Override
    public McpServerInfo serverInfo() {
        return serverInfo;
    }

    @Override
    public void close() throws Exception {
        if (process != null) {
            process.destroy();
        }
    }

    // -------------------------------------------------------------------------

    private JsonRpcResponse sendRequest(String method, JsonNode params) throws Exception {
        lock.lock();
        try {
            long id = nextId.incrementAndGet();
            JsonRpcRequest req = JsonRpcRequest.of(id, method, params);
            String json = mapper.writeValueAsString(req);
            writer.write(json);
            writer.newLine();
            writer.flush();
            String line = reader.readLine();
            if (line == null) {
                throw new RuntimeException("mcp: server closed stdout");
            }
            return mapper.readValue(line, JsonRpcResponse.class);
        } finally {
            lock.unlock();
        }
    }
}
